### ISSUE TITLE
Dependency resolution fails on transitive renaming

### ISSUE THREAD

**Issue by gagbo** [rust-lang/cargo#10447]
### Problem

I don't understand why I can't have "package renaming transitivity" in my use case. When using crate renaming in a project ("multi-bars") that is used as a dependency of a bigger one, the package rename within `multi-bars` is ignored.

### Steps

In `multi-bars/Cargo.toml` I have
```toml
bar7 = { version = "7", registry = "private", package = "bar" }
```
and everything works fine (I can compile and publish `multi-bars`, since `bar v7` is available)

In the project `foo` that uses `multi-bars`:
```toml
multi-bars = { version = "0.1.1", registry = "private" }
```
cargo cannot figure out the dependency tree : 
```
$ cargo version
cargo 1.59.0 (49d8809dc 2022-02-10)

$ cargo tree
    Updating crates.io index
    Updating `private` index
error: no matching package named `bar7` found
location searched: registry `private`
required by package `multi_bars v0.1.1 (registry `private`)`
    ... which satisfies dependency `multi-bars = "^0.1.1"` of package `foo v3.1.2 (/path/to/foo)`
```
I don't understand why cargo can't figure out the dependency I want to use. Is there something obvious I missed ?

### Possible Solution(s)

I don't understand why Cargo knows the name of the package to look for in `multi-bars` when used directly but not when used indirectly.

### Notes

- I also posted on [users.rust-lang](https://users.rust-lang.org/t/transitivity-issues-in-crate-renaming-in-private-registry/72467), but as I'm trying to make sense out of this I'm slowly starting to think I did nothing wrong, so I'm posting here. Feel free to close the issue if that's wrong
- I didn't see the issue when working locally (with path overrides), but only once I stopped using the overrides and started using the registry index. (And cargo tree works on `multi-bars` even without path overrides)
- Not having the issues with overrides makes reproduction a little tricky, since I can't just make 2 repos or 1 monorepo to trigger it. If it's really necessary I suppose I can create a few dummy crates on crates.io, if it's necessary tell me and I'll make some time later to do it; I'm trying to postpone this to avoid unnecessary pollution

### Version

```text
cargo 1.59.0 (49d8809dc 2022-02-10)
release: 1.59.0
commit-hash: 49d8809dc2d3e6e0d5ec634fcf26d8e2aab67130
commit-date: 2022-02-10
host: x86_64-unknown-linux-gnu
libgit2: 1.3.0 (sys:0.13.23 vendored)
libcurl: 7.80.0-DEV (sys:0.4.51+curl-7.80.0 vendored ssl:OpenSSL/1.1.1l)
os: OracleLinux 35.0.0 [64-bit]
```
```

**Comment by gagbo** [comment]
I was told to check the dep entry in my registry implementation (ktra), and here are my findings:

I've checked the dependency entry in my registry, and there's definitely an error there but I'm not sure it should have an impact:

```
{
      "name": "bar",
      "req": ">=7.1.0, <8.0.0",
      "features": [],
      "optional": false,
      "default_features": true,
      "target": null,
      "kind": "normal",
      "registry": null,
      "package": "bar7"
    },
```

From looking at the Cargo book it seems the `package` key should be a `explicit_name_in_toml` key, but anyway, cargo should use the `name` to search for the package right ?

**Comment by gagbo** [comment]
I got it wrong, the Cargo book extract I looked at was the payload used in publish, not the expected format at rest, so there's indeed an issue in my registry. Sorry for bothering, thanks for rubberducking.
